python/pyspark/streaming/util.py

```diff
@@ -64,7 +64,11 @@ def call(self, milliseconds, jrdds):
             t = datetime.fromtimestamp(milliseconds / 1000.0)
             r = self.func(t, *rdds)
             if r:
-                return r._jrdd
+                # Here, we work around to ensure `_jrdd` is `JavaRDD` by wrapping it by `map`.
+                # org.apache.spark.streaming.api.python.PythonTransformFunction requires to return
+                # `JavaRDD`; however, this could be `JavaPairRDD` by some APIs, for example, `zip`.
+                # See SPARK-17756.
+                return r.map(lambda x: x)._jrdd
```

Comment:
wouldnt introducing this additional map cause a slight perf regression?

Reply:
Yup, it guess it would but I could not think of a better and simpler way (and I saw this workaround is already used above in `python/pyspark/streaming/context.py`). If a little bit more uglier (possibly flaky imho) way is okay for this concern, I could do this, for example, as below:

```python
if r._jrdd.getClass().toString() == "JavaRDD":
    return r._jrdd
else:
    return r.map(lambda x: x)._jrdd
```

Reply:
@tdas, I believe this one is something we should fix ... would you like me to keep this way, try the way above or find another way to avoid a map even if it is complex? 

Otherwise, I can run a benchmark on RDD with extra map too see the slight perf diff (although I am not sure yet if I can show the meaningful diff).

Reply:
@tdas, it's fine if you feel we should find another way or make a small benchmark. I can try. But just wonder which one you'd prefer before I go too far in a different way you'd prefer.

Reply:
Personally, I think the only applying the `map` when the result is not JavaRDD is a good incremental improvement (since otherwise the code path fails right?).

Reply:
Thanks for review @holdenk. Let me push the change.